### DISCUSSION TITLE
image vulnerabilities

### DISCUSSION THREAD

**Discussion by kl4072** [apache/airflow#37608] [Q&A] [answered]
Hello Community, I have been able to extend official airflow images to pass through twistlock scan's however there is an issue related to go lang 1.20.7 which i am unable to solve, does anyone have any idea which reference file maybe causing this as a false positive?

I have got exception from secops to use image with the false positives in lower environments and the approval process is tedious, so
looking to see if i can customize the image. If anyone has run into this please chime in how you resolved.

**Comment by potiuk** [comment]
https://airflow.apache.org/docs/docker-stack/index.html#fixing-images-at-release-time describes what to do in this case. Generally you have three options:

1) Build your own custom image following the examples we share there - using latest base image and possibly bumping dependencies you want to bump. There are quite a few examples there when you follow the links.

2) wait for new version of Airflow and upgrade to it. Airlfow images are updated to latest "non-conflicting" dependencies and use latest "base" image at release time, so what you have in the reference images at the moment we publish the image / release the version is what is "latest and greatest" available at the moment with the base platform we use (debian bookworm is the reference image we use) We are currently voting on Airlfow 2.8.2 RC1 as you can check the devlist announcement (find how to subscribe to and read archives of devlist here https://airflow.apache.org/community/) and test it and report feedback in #37617 and report results of your testing there. You will find images with 2.8.2rc1 here https://hub.docker.com/r/apache/airflow/tags - they are built **today** and see if your scan results are solved there.

2) If the base platform we use (currently debian-bookworm) does not contain the lates versions you want and you want to use other base images, you can take a look at what system dependencies are installed in the latest Doickerfile of airflow and take inspiratio from it and build your own image (or copy the dockerfile and modify it as you see fit) https://github.com/apache/airflow/blob/main/Dockerfile

This gives you all the flexibility you need. You can rely on Airlow bumping to latest, non-conflicting 3rd-party dependency versions regularly - with every release, so one of the strategies you can take to keep your Airflow secure is to engage in early testing of RC candidates and upgrading to newer airlfow releases as soon as they happen.

Also if you think that something is false positive - we encourage you to contribute back by analysing such  scan reports and trying to come up with exploitation scenarios - and report it here for other people who migh be looking at it. If you have an exploitation scenario you can report it privately following our security policy https://github.com/apache/airflow/security/policy - this email is there if you will find a way how to exploit third-party dependency vulnerability in airflow, and all the volunteer maintainers highly appreciate efforts of our commercial users to report such exploitable vulnerabilities (after analysing them) as a way to contribute back for the community that makes the free software available for you at no charge (and no guarantees and no support but that's another story).

**Reply by kl4072** [reply]
Thanks for the response, I was able to build using the docker file upgrade the dependent python packages, that were causing issues and create a new image. The main vulnerabilities are gone. 

However, there is a false positive related to [CVE-2023-39323](https://nvd.nist.gov/vuln/detail/CVE-2023-39323) will need to look into which package/tar.gz is causing this issue. go lang issues which I consider false positives have been there for the past 3 months, i have run these images against our scanner and is getting reported on twistlock. 

Will try and break down the dockerfile and see what can be done, Thanks for your videos on Youtube, they were really helpful in understanding some of the build internals. I don't fully know all the details yet but looking to learn and contribute to the community when possible.

**Reply by potiuk** [reply]
In the future if you have similar issue just mark it as such (false positive). 

It's never possible to remove all vulnerabilities and if you believe it is false positive - mark is as such in your sstems. If your security team does not allow you to do so, then they are doing it wrong. It's perfectly normal that there are some vulnerabilities in the image that cannot be exploited - becasuse a) some tooling is uing it b) some dependencies have it c) some binaries (like in this case most likely) have been built with older go version. And they are mostly coming from base images you build it on. In this case - this is vulnerability that **likely** comes from the debian-bookworm  - and they have not yet addressed it.

You can  

a) find and remove those binaries that have it (good luck)
b) use different base image 
c) review and mark this vulnerability as "known, not affecting us" for your security team. Actually you should even ask YOUR security team to verify that . This is generally the easiest way if you can check that in-fact what your security scan shows somethigni that is never used or that it shows it wrongly.

And YOU  shoudl be able to do it because YOUR security scan should be able to tell you what exactly is causing them to think there is the vulnerability. It's also an optuion that YOUR security scan is showing false positive and there is no security issue at all. Security scanners are known to detect things which do not exist. So it's really in YOUR hands to decide what to do. Removing the CVE by removing affected binaries is only one of  ways of dealing with the problem and you seem to be desperately following only that one path.

**Comment by kl4072** [comment, marked as answer]
I was able to customize image and fix all the vulnerabilities my main issue was with the DOCKER_CLI_VERSION=24.0.6 increment this to 25.0.3 rest is easy. It was a painful process finding out the rot cause but learned a lot in the process.

Anyone running into go lang vulnerabilities this is it!

**Reply by potiuk** [reply]
Thanks. Actually I was looking at some of that recently and yes I think we will bump all the docker related tooling in the image, that was a good find, thanks for that.

**Reply by potiuk** [reply]
Yeah, thanks for that again (sorry for being a bit pushy on security@, but we really cannot do everything with generic security scans. But in this case yes, we can improve it by simply not installing specific version (we bump it from time to time, and we used it mostly to see if we are not using any newer features accidentally, but in this case I think yes - updating to latest docker binary CLI is a good call (and I will change it in our future images).

**Reply by kl4072** [reply]
No worries! Thanks for all that the community does glad to help a tiny bit. It was not easy to find, but understanding the build process was key.

**Reply by potiuk** [reply]
https://github.com/apache/airflow/pull/37651 - merged
